### DISCUSSION TITLE
Password Protect Individual entries on the Menu

### DISCUSSION THREAD

**Discussion by jamaya77** [ipxe/ipxe#346] [Q&A] [answered]
Hi,

I was wondering if anyone has any examples or suggestions as to how I can password protect the individual entries on the iPXE menu. Below is an example of how I'm trying to set it up, but the example below prompts for the username and password, but it will accept any credentials or no credentials at the login prompt. Thanks in advance.

![ipxepass](https://user-images.githubusercontent.com/80473349/116638159-b335b600-a933-11eb-8817-4213c4b143e3.png)

``#!ipxe

set user admin
set pass test1234

menu iPXE Boot Menu
item --gap System Center Configuration Manager: 
item sccm_boot Boot to SCCM
item

choose target && goto ${target}

#
#sccm
#
:sccm_boot
login
chain http://${username:user}:${password:pass}@192.168.140.21/sccm/sccmbios.ipxe
#``

**Comment by NiKiZe** [comment, marked as answer]
login is only used to populate the data, the validation should be done by the http server, and any failure in the chain comman needs to be handled.

**Reply by NiKiZe** [reply]
Documentation at https://ipxe.org/cmd/login
finalize your chain line with || goto sccm_boot
to your login line, add || goto menu
which would allow ctrl-c at login to return to the menu (label needs to be added to the menu as well)

**Comment by sebaxakerhtc** [comment]
Why to set user and password?
Leave just login and type in credentials. 
You already set user and pass. Doesn't matter, what to type ind after.
Try to remove this lines:
```
set user admin
set pass test1234
```
And it should work like this:

```
:sccm_boot
login
chain http://${username}:${password}@192.168.140.21/sccm/sccmbios.ipxe || goto failed
```
This is my login menu entry you can use as example:

```
:myiso
login || goto failed
imgfree
sanboot http://example.com/${username}bs/img/${password}.iso || goto failed
```

**Reply by JohnLar42** [reply]
If you can setup authentication on your http/https server you can do this.
```
login
menu Boot Menu
item public Public Item
imgfetch http://${username:uriencode}:${password:uriencode}@server/directory/smallcheckfile && item scm_boot hidden item
imgfetch http://${username:uriencode}:${password:uriencode}@server/directory2/smallcheckfile && item scm_boot hidden item2
item public2 Public Item2
```

**Comment by jamaya77** [comment]
I must be  missing something. Right now my setup is a CentOS 8 server, running apache 2.4.
Do I need to configure apache so its protected with a username and password OR change the permission of a file (or folder) to a specific user and password.

When use the above configuration, any username and password will authenticate.

Thanks for your time in advance.

**Reply by mcb30** [reply]
The purpose of the `login` command (as documented at https://ipxe.org/cmd/login) is merely to allow the user to enter a username and password.  What you then choose to do with the entered username and password is entirely up to you.

In your example, you have
```
login
chain http://${username}:${password}@192.168.140.21/sccm/sccmbios.ipxe || goto failed
```

You are therefore telling iPXE to allow the user to enter a username and password, and you are then using this username and password as HTTP authentication to your web server 192.168.40.21.

You will need to configure your web server to require authentication for the URL that you are attempting to access, and to validate the username and password against whatever authentication database you want to use.  You can find some documentation of doing this for the Apache web server at https://httpd.apache.org/docs/2.4/howto/auth.html#gettingitworking

**Reply by sbrudenell** [reply]
what if I want to password protect access to the ipxe `shell`?

**Reply by JohnLar42** [reply]
> what if I want to password protect access to the ipxe `shell`?

You can do an auth check with shell as the success of the auth check. && is AND, it only runs in success

:shell
login
imgfetch http://${username:uriencode}:${password:uriencode}@server/directory/smallcheckfile && shell || goto failure

::failure
(handle failure)
